components/callbacks/chasm_invocation.go

```diff
@@ -0,0 +1,73 @@
+package callbacks
+
+import (
+	"context"
+	"encoding/base64"
+	"errors"
+	"fmt"
+
+	"github.com/nexus-rpc/sdk-go/nexus"
+	"go.temporal.io/api/serviceerror"
+	persistencespb "go.temporal.io/server/api/persistence/v1"
+	"go.temporal.io/server/chasm"
+	"go.temporal.io/server/common/namespace"
+	"go.temporal.io/server/service/history/queues"
+)
+
+type chasmInvocation struct {
+	nexus      *persistencespb.Callback_Nexus
+	attempt    int32
+	completion nexus.OperationCompletion
+}
+
+var ErrUnimplementedHandler = serviceerror.NewUnimplemented("component does not implement NexusCompletionHandler")
+
+func (c chasmInvocation) WrapError(result invocationResult, err error) error {
+	if failure, ok := result.(invocationResultFail); ok {
+		return queues.NewUnprocessableTaskError(failure.err.Error())
+	}
+
+	if retry, ok := result.(invocationResultRetry); ok {
+		return queues.NewDestinationDownError(retry.err.Error(), err)
+	}
+
+	return err
+}
+
+func (c chasmInvocation) Invoke(ctx context.Context, ns *namespace.Namespace, e taskExecutor, task InvocationTask) invocationResult {
+	// Get back the base64-encoded ComponentRef from the header.
+	encodedRef, ok := c.nexus.GetHeader()[chasm.NexusComponentRefHeader]
+	if !ok {
+		return invocationResultFail{errors.New("callback missing CHASM header")}
+	}
+
+	decodedRef, err := base64.RawURLEncoding.DecodeString(encodedRef)
+	if err != nil {
+		return invocationResultFail{fmt.Errorf("failed to decode CHASM ComponentRef: %v", err)}
+	}
+
+	ref, err := chasm.DeserializeComponentRef(decodedRef)
+	if err != nil {
+		return invocationResultFail{fmt.Errorf("failed to deserialize CHASM ComponentRef: %v", err)}
+	}
+
+	// Attempt to access the component and call its invocation method. We execute
+	// this similarly as we would a pure task (holding an exclusive lock), as the
+	// assumption is that the accessed component will be recording (or generating a
+	// task) based on this result.
+	_, err = e.ChasmEngine.UpdateComponent(ctx, ref, func(ctx chasm.MutableContext, component chasm.Component) error {
```

Comment:
This doesn't do an RPC, you're going to have to add an RPC to the shard that owns the execution of the original caller.
We can then optimize it so if the ref is something that we can resolve locally we would bypass the RPC call.

Reply:
The RPC would be `historyservice.CompleteNexusOperation`, you'll want to update that method to invoke `ChasmEngine.UpdateComponent` and take a different form of completion.

Feel free to rename the `completion` field to `hsm_completion` to disambiguate between HSM and CHASM refs.

Reply:
Let's also not change the Nexus RPC stuff for now since we are not using that SDK for delivering the completions.

We can change it later when the SDK supports pluggable transports if we want to but that's probably not critical.